src/controller/src/lib.rs

```diff
@@ -174,17 +176,18 @@ pub struct Controller<T = mz_repr::Timestamp> {
     /// Arguments for secrets readers.
     secrets_args: SecretsReaderCliArgs,
 
-    /// A map associating a global ID to a vector of all the watch sets
-    /// that that ID is part of which have not yet been fulfilled for that object.
+    /// A map associating a global ID to the set of all the unfulfilled watch
+    /// set ids that include it.
     ///
     /// See [`self.install_watch_set`] for a description of watch sets.
     // When a watch set is fulfilled for a given object (that is, when
     // the object's frontier advances to at least the watch set's
-    // timestamp), the corresponding entry will be removed from the
-    // vector here. That way, when the entire watch set is fulfilled,
-    // the corresponding `Rc` will be the last reference to it, and
-    // the call to `Rc::into_inner` will succeed.
-    objects_to_unfulfilled_watch_sets: BTreeMap<GlobalId, Vec<Rc<(T, Box<dyn Any>)>>>,
+    // timestamp), the corresponding entry will be removed from the set.
+    unfulfilled_watch_sets_by_object: BTreeMap<GlobalId, BTreeSet<WatchSetId>>,
+    /// A map of installed watch sets indexed by id.
+    unfulfilled_watch_sets: BTreeMap<WatchSetId, (BTreeSet<GlobalId>, T, Box<dyn Any>)>,
+    /// A sequence of numbers used to mint unique WatchSetIds.
+    watch_set_id_seq: RangeFrom<u64>,
```

Comment:
Nit: I think this could also use `mz_ore::id_gen::IdGen`, if you add a `next_id` method for the check in `uninstall_watch_set`. That would avoid an `unwrap` below and slightly improve readability.

Reply:
ah, nice! I didn't know we had this utility

Reply:
Switched to the `Gen` type but I didn't add the `next_id` value. It doesn't provide a lot of value since the only way to get a valid id instance is to mint it (well, somebody could convert a u64 to an id but that's not really a concern)